### DISCUSSION TITLE
Flows: Improve the Log Activity Panel

### DISCUSSION THREAD

**Discussion by erondpowell** [directus/directus#15870] [Feature Requests]
Flows is totally cool! But the log panel has (in my opinion) UI/UX issues.
I find myself immediately reaching for Postman to debug/setup just about any flow.

In this discussion, I'm going to address what I consider to be the biggest pain points with the Log panel.
Then at the end, suggest an alternative *(wireframe/rough-draft)* UI layout to try to address these pain points. 

---

### The activity Log menus are not a 1:1 mapping to the flow object

When you toggle open `$trigger`, it has `$accountability` nested in to it.

<img width="308" alt="Screen Shot 2022-10-03 at 11 52 36 AM" src="https://user-images.githubusercontent.com/62306630/193634246-904eeaab-c09f-4b2b-9772-7dd67e0b0c6a.png">

In reality, `$accountability` is actually a sibling key on the Flow Object.

*(note: Since `options` is nested in the Data Studio log panel, at a glance it seems like it exists on the flow object, though it doesn't.)*

![CleanShot 2022-10-03 at 11 52 54](https://user-images.githubusercontent.com/62306630/193634731-a338958b-e454-474f-879f-3413e5d41084.png)

On a similar note, `$last` is not provided within the logger at all. I'd guess the justification here is that `$last` is not a "distinct value", but exists for convenience.... I think this would be nice to see in the Logger as an additional sanity check when configuring each operation in the flow.

---

### The activity Log menus are not a 1:1 mapping to the flow object... (Part 2)

This is quite similar to the first issue. Consider this example below.

By looking at the Log, we can't quickly identify the `key.subkey1.subkey2` that we need.

Again, it feels like you should use `$trigger.accountability.user`, but we actually need to use `$accountability.user`.
At a glance, it feels like you should use `abc.payload.key` but we actually need to use `abc.key`.

![CleanShot 2022-10-03 at 13 10 22@2x](https://user-images.githubusercontent.com/62306630/193648346-d9b7bc91-6e1f-4e48-80af-8fa5f89bc938.png)

I think this mismatch likely leads to a lot of confusion we've seen around accessing a desired value.

---

### Redundant and divergent naming conventions
We use the same name for different things and different names for the same *(well, very very similar)* things.
This makes it harder to parse through the info and be certain of what you're looking at.

In the Data Studio's Log panel, we use `payload` to label the value appended under a trigger or operation key. 

<img width="321" alt="CleanShot 2022-10-03 at 12 08 52@2x" src="https://user-images.githubusercontent.com/62306630/193637330-cb493a78-12a0-4456-b36f-6748c70f35a9.png">

In an Event Hook trigger configured to trigger on `items.create` events, the body of the post request is labeled `payload`.

<img width="1082" alt="Screen Shot 2022-10-03 at 11 49 23 AM" src="https://user-images.githubusercontent.com/62306630/193633654-063b5683-a284-485b-a3cb-b2b5cc4628d5.png">

In a Webhook trigger configured to trigger on `POST` requests, the body is called `body`.

![CleanShot 2022-10-03 at 12 29 35](https://user-images.githubusercontent.com/62306630/193641252-c2dad337-5aca-4289-9337-a6291fc27b93.png)

There may be additional overlapping names, but this was the most prominent. 

---

### JSON object wrapping not effectively handled

https://user-images.githubusercontent.com/62306630/193662898-d7963303-7812-4dd3-9fd3-bdc028ce9be3.mp4

We should add a scroll bar here.

---

### $all exists on object but isn't accessible by users

Right now, `Response Body` uses $all variable... 

<img width="385" alt="Screen Shot 2022-10-11 at 4 42 07 PM" src="https://user-images.githubusercontent.com/62306630/195203750-3b3c8e4d-f4a3-48b4-8fb4-7be8a39ae773.png">

but this isn't available for end-users. which is somewhat confusing.

---

### There's no option to delete/cleanup logs

<img width="1133" alt="Screen Shot 2022-10-11 at 5 15 40 PM" src="https://user-images.githubusercontent.com/62306630/195208221-91ab45ff-f7d1-41cf-8bc1-0bedcd7e47b5.png">

This takes up memory and gets hard to sort through. There should be several delete options built-in:
- individual/batch
- delete all
- schedule

---

### New UI Suggestion for Logs Panel

The following wireframed layout includes the general suggestions/thoughts I have to improve the log panel.

![Work Journal 2-62](https://user-images.githubusercontent.com/62306630/193660687-55d43d3b-73ae-4684-8451-2532055c0b09.jpg)

- In the Log panel, the Flow Object should look like an object, with the option to toggle display the `value` under each `key`.
- Users should be able to see the entire Flow Object "exactly as it is", with a clear 1:1 mapping between `key:value`. It also reduces confusion between Config Options (input) and data appended to the Flow Object (output).
- When data is long, it wraps and becomes very hard to read. Scroll bar fixes this.

Again, this is a wireframe, not a fully flushed out solution.
In the process of clarifying the Flow Object, several other design questions are raised:

- The location of the console log feels a bit weird in my wireframed layout.
- Perhaps all the options should go before Flow Object, or just get nested slightly differently.
- ????

Curious to see how others agree/disagree/tangent on my thoughts 🤠

**Comment by br41nslug** [comment]
I agree the way the current flow context object represented in the UI is confusing as it doesn't map intuitively to the object actually used for mustache templates. I like the idea of splitting up `payload` data and `options` combined with a closer relation to the actual object would save a lot trial and error. We could even add a "view raw" option like we have for fields which would just dump the entire object in a `pre`.

Not completely related but I had an additional UI/UX idea while going through this is that it would be great to have a mustache testing widget here in the sidebar. This simple widget would effectively just be an input box and result panel where you'd input the mustache template and it would show you the resulting output based on the last logged object (or an error if invalid/non-existent).
![image](https://user-images.githubusercontent.com/9389634/194581498-94fbb23c-00a5-45d3-b4ce-bdffb14ab2ac.png) ![image](https://user-images.githubusercontent.com/9389634/194581547-b518d9d1-ab18-4bb5-87af-8e6864ca8c08.png)

**Reply by erondpowell** [reply]
Tim, that var tester would be an insanely nice sanity check.... *Soooo many* console logs would be saved.

**Reply by erondpowell** [reply]
> I like the idea of splitting up payload data and options combined with a closer relation to the actual object would save a lot trial and error.

Yeah. Especially for users new to JSON, I bet the logger not being a 1:1 mapping is big a struggling point.
They need to be able "to see the whole thing" in order to know where the data is and to index it properly.

**Reply by zakwear** [reply]
Variable tester would be wonderful. 

Could also embed the variable tester in an operation widget, and have it check against the most recent logged values.

I agree the abbreviated syntax is more confusing than just typing, "$last.payload[0].id". Access to a separated .options would be great and also chop down on the number of operations.

**Comment by erondpowell** [comment]
@directus/team

**Comment by josephdpurcell** [comment]
I vehemently agree with points:

* The activity Log menus are not a 1:1 mapping to the flow object
* The activity Log menus are not a 1:1 mapping to the flow object... (Part 2)

If the Log was 1:1 with the data chain it would have saved me (subjectively speaking) a lot of time, see also https://github.com/directus/directus/issues/16674. Specifically, I struggled with the Read Data op knowing what to put for the UUID, which was `{{ $trigger.body.keys[0] }}`. If I understand the `New UI Suggestion for Logs Panel` proposal correctly, I could have simply expanded `$trigger` menu and seen those values and come to the correct conclusion. Instead, I saw `Trigger` and there was no trigger in my brain that `Trigger` == `$trigger` (pun intended). It's comical to me now that I missed something so obvious. And, now that I know how the data chain is structured I won't have this problem again. But, for the next user like me, they would probably benefit from simply changing the `Trigger` label to `$trigger` at minimum. But, more than that, I think the UI recommendations to better make the UI map 1:1 with the data chain structure are great.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by phazonoverload** [comment]
This request has continued to come up, so on February 8 we're inviting the community to come and have a chat with us about this feature request at our Request Review in our Discord server. This will help us have some dedicated time to talk about what implementation would have the most sense. Please feel free to join us as https://directus.chat/ (the event is already listed and you can set a reminder)

**Comment by ched-dev** [comment]
I would love to see options to clear out logs, and the ability to "lock" a log so I can save specific ones while still being able to clean up all.

**Reply by mighty-hq** [reply]
would love to be able to refresh logs in flow without having to refresh page

**Comment by mighty-hq** [comment]
Idea to improve the life of devs  :)) Would be great if we can add a note to each step in flows. These would be used only internally
 obviously and would help with development. 


![ff](https://github.com/directus/directus/assets/81202462/42cb2f81-c437-4462-af1e-64618ecab0d6)

This is an example of how its done in another platforms flows.

![image](https://github.com/directus/directus/assets/81202462/e879c9be-e001-4ac5-942a-961c12855741)

**Reply by MarkBurvs** [reply]
I would love this. I've got so many flows going on in my project now, it often takes me a bit of digging to remember what they're all for. Especially when you get into using some flows as loops triggered by other flows, it can all get complicated fast!

**Comment by mighty-hq** [comment]
2 More ideas, and  these ones are actual problems. 

1.When editing flows, you click edit, and the Edit Operation Side Menu shows up, well that immediately resets  the view of the drag and drop interface to the top, so when you close that Side Menu, you are reset to the top position. When having larger flows that stack, that is really annoying, because you have to pan and search again for the flow step .


2. Ability to zoom in/our on the flows, this would really help

**Comment by mario4rma** [comment]
Just read the whole discussion and I really like these UI and DX improvements.
Just got an additional idea:
What if we would have a „copy“ button next to every item in the logs.
Expand trigger/payload an you see key.subkey1.subkey2 and you just click the little copy icon next to it, and that’s it.
